show-mod.php

```diff
@@ -1,185 +1,345 @@
+	foreach ($tagdata as $tagrow) {
+		$row = explode(",", $tagrow);
+		$tags[] = array('name' => $row[0], 'color' => $row[1], 'tagid' => $row[2], 'text' => $alltags[$row[2]]);
+	}
+}
+
+$view->assign("tags", $tags);
+
+$releases = $con->getAll("
+	select 
+		`release`.*,
+		asset.*
+	from 
+		`release` 
+		join asset on (asset.assetid = `release`.assetid)
+	where modid=?
+	order by release.created desc
+", array($asset['modid']));
+
+foreach ($releases as &$release) {
+	$compatibleGameVersions = array();
+	if (!empty($compatibleGameVersionsCached = trim($release["tagscached"]))) {
+		$tagdata = explode("\r\n", $compatibleGameVersionsCached);
 		foreach ($tagdata as $tagrow) {
 			$row = explode(",", $tagrow);
-			$tags[] = array('name' => $row[0], 'color' => $row[1], 'tagid' => $row[2], 'text' => $alltags[$row[2]]);
+			$compatibleGameVersions[] = array('name' => $row[0], 'color' => $row[1], 'tagid' => $row[2]);
 		}
 	}
+	if (count($compatibleGameVersions)) {
+		usort($compatibleGameVersions, 'rcmpVersionTag');
+		$release['highestver'] = $compatibleGameVersions[count($compatibleGameVersions) - 1]['name'];
+	} else {
+		$release['highestver'] = "";
+	}
 
-	$view->assign("tags", $tags);
-
-	$releases = $con->getAll("
-		select 
-			`release`.*,
-			asset.*
-		from 
-			`release` 
-			join asset on (asset.assetid = `release`.assetid)
-		where modid=?
-		order by release.created desc
-	", array($asset['modid']));
+	$release['isPreRelease'] = isPreRelease($release['modversion']);
+	$release['compatibleGameVersionIds'] = array_map(fn($t) => intval($t['tagid']), $compatibleGameVersions);
+	$release['compatibleGameVersions'] = groupMinorVersionTags($compatibleGameVersions);
+	$release['file'] = $con->getRow("select * from file where assetid=? limit 1", array($release['assetid']));
+}
+unset($release);
+
+usort($releases, "cmpReleases");
+$releases = array_reverse($releases);
+
+
+/*
+	Determine the game versions of interest:
+		stable and
+		unstable (newer than current stable if any)
+
+	Match mod releases:
+		latest stable release that is for the stable version of the game -> recommend
+		latest unstable version that is either
+			for the unstable version of the game, or
+			for the stable version of the game, if the release is a newer unstable version than the stable release (only if there is no release for the unstable game version)
+		-> recommend for testers
+		If there are not releases for for either of these, select the latest release -> latest release for an outdated version of the game
+
+	Examples assuming current game version = 5, an a newer unstable version = 5p, GV = game version, RV = mod release version:
+		GV  RV
+		5   2.5
+		5   3
+		5   4.1  -> Recommended
+		5p  4.2  -> For testers
+
+		5   2.5
+		5   3    -> Recommended
+		5   4.p1
+		5   4.p2 -> For testers
+
+		5   2.5
+		5   3    -> Recommended
+		5   4.p1
+		5p  4.p2 -> For testers
+
+		2   2.5
+		2   3
+		3   4.p1
+		3   4.p2 -> Latest outdated

```

Comment:
assuming that 4.p2 is for GV 2

Reply:
No, 4.p2 is for GV 3, but the latest GV for this example would be 5, so there is no up to date version of the mod